for loop over mutable elements in costume iterator that returns an array of strings

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: windows, Microsoft Windows 10 Pro v19042 64-bit
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i7-6600U CPU @ 2.60GHz
CC version: Error: exec failed (CreateProcess) with code 2: The system cannot find the file specified.

 cmd: cc --version

getwd: C:\Users\UK-Lap024d\Documents\v\rep_error
vmodules: C:\Users\UK-Lap024d\.vmodules
vroot: C:\Users\UK-Lap024d\Documents\v_store\v
vexe: C:\Users\UK-Lap024d\Documents\v_store\v\v.exe
vexe mtime: 2021-11-01 17:33:39
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 dcf230c

Git version: git version 2.33.1.windows.1
Git vroot status: weekly.2021.43-61-gdcf230ca
.git/config present: true
thirdparty/tcc status: thirdparty-windows-amd64 0f5168cd
```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg founderror.v`
```v
struct Struct{
mut:
	array [][]int
}
fn (s Struct) rows() StructsRowIterator{
	return StructsRowIterator{
		array: s.array
		position: 0
	}
}

struct StructsRowIterator{
	Struct
mut:
	position int
}
fn (mut s StructsRowIterator) next() ?[]int{
	if s.position>=s.array.len{
		return error("out of range")
	}
	defer{
		s.position++
	}
	return s.array[s.position]
}


mut s := Struct{
	array:[[1,2,3],[4,5,6]]
}
println(s)
mut si := s.rows()
println(si)
for mut row in si{
	println(row)
}

```

**What did you expect to see?**

printing out of the array of strings, works without the 'mut' from the 'mut row in' but not with mut


**What did you see instead?**
```
==================
C:/Users/UK-LAP~1/AppData/Local/Temp/v/founderror.6241902178958351573.tmp.c:6642: warning: cast between pointer and integer of different size
C:/Users/UK-LAP~1/AppData/Local/Temp/v/founderror.6241902178958351573.tmp.c:10891: error: pointer expected
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2165,7 +2165,11 @@ fn (mut g Gen) for_in_stmt(node ast.ForInStmt) {
 		g.writeln('\tif (${t_var}.state != 0) break;')
 		val := if node.val_var in ['', '_'] { g.new_tmp_var() } else { node.val_var }
 		val_styp := g.typ(node.val_type)
-		g.writeln('\t$val_styp $val = *($val_styp*)${t_var}.data;')
+		if node.val_is_mut {
+			g.writeln('\t$val_styp* $val = ($val_styp*)${t_var}.data;')
+		} else {
+			g.writeln('\t$val_styp $val = *($val_styp*)${t_var}.data;')
+		}
 	} else {
 		typ_str := g.table.type_to_str(node.cond_type)
 		g.error('for in: unhandled symbol `$node.cond` of type `$typ_str`', node.pos)

--- a/vlib/v/tests/for_in_mut_struct_val_test.v
+++ b/vlib/v/tests/for_in_mut_struct_val_test.v
@@ -1,0 +1,45 @@
+struct Struct {
+mut:
+	array [][]int
+}
+
+fn (s Struct) rows() StructsRowIterator {
+	return StructsRowIterator{
+		array: s.array
+		position: 0
+	}
+}
+
+struct StructsRowIterator {
+	Struct
+mut:
+	position int
+}
+
+fn (mut s StructsRowIterator) next() ?[]int {
+	if s.position >= s.array.len {
+		return error('out of range')
+	}
+	defer {
+		s.position++
+	}
+	return s.array[s.position]
+}
+
+fn test_for_in_mut_struct_val() {
+	mut s := Struct{
+		array: [[1, 2, 3], [4, 5, 6]]
+	}
+	println(s)
+	mut si := s.rows()
+	println(si)
+
+	mut rets := []string{}
+	for mut row in si {
+		println(row)
+		rets << '$row'
+	}
+	assert rets.len == 2
+	assert rets[0] == '[1, 2, 3]'
+	assert rets[1] == '[4, 5, 6]'
+}

